src/model.js

```diff
@@ -0,0 +1,187 @@
+    const { geometry: { spatialReference: { wkid } = {} } } = requestQuery;
+    const queryWkid = wkid || requestQuery.inSR;
+
+    // assume 4326 if no SR is provided
+    if (!queryWkid) {
+      return 4326;
+    }
+
+    // 102100 is the old Esri code for 3857 but not recognized for proj4
+    if (queryWkid === 102100) {
+      return 3857;
+    }
+    // If the input wkid is one of the set known to proj4, return it in an object
+    if (PROJ4_WKIDS.includes(queryWkid)) {
+      return queryWkid;
+    }
+
+    logger.debug(`unsupported wkid ${queryWkid} provided`);
+  }
+
+  isSupportedGeometry(requestQuery) {
+    return (
+      requestQuery.geometry &&
+      requestQuery.geometryType === 'esriGeometryEnvelope'
+    );
+  }
+
+  getExtent(geometry) {
+    if (!objTypeGeometryQuerySchema.validate(geometry).error) {
+      const { xmin, ymin, xmax, ymax } = geometry;
+      return [[xmin, ymin], [xmax, ymax]];
+    }
+
+    const extent = geometry.split(',').map(Number);
+    if (!arrayTypeGeometryQuerySchema.validate(extent).error) {
+      return [extent.splice(0, 2), extent];
+    }
+  }
+
+  addSortOptionsToPortalQuery(portalQuery, orderByFields) {
+    const { orderBy } = orderByFields;
+    const orderByArr = orderBy.split(' ');
+    portalQuery.sortField = orderByArr[0] || 'title';
+    portalQuery.sortOrder = orderByArr[1] || 'DESC';
+    return portalQuery;
+  }
+
+  addBboxToPortalQuery(portalQuery, requestQuery) {
+    const extent = this.getExtent(requestQuery.geometry);
+    const inSR = extent && this.getSpatialReference(requestQuery);
+    if (extent && inSR) {
+      const normalizedExtent = convertToWGS84(extent, inSR);
+      portalQuery.bbox = normalizedExtent.join(',');
+    }
+    return portalQuery;
+  }
+
+  async getPortalItems(portalUrl, query, maxPageSize) {
+    const firstPage = await this.fetchItemsFromPortal(portalUrl, query);
+    const totalBatch = this.getTotalBatch(firstPage.total, maxPageSize);
+    if (totalBatch === 1) {
+      return {
+        items: firstPage.results,
+        count: firstPage.total
+      };
+    }
+    const remainingRequests = this.buildRemainingPageRequests({ portalUrl, query, totalBatch, maxPageSize });
+    const remainingItems = await this.getRemainingPortalItems(remainingRequests);
+    return {
+      items: [...firstPage.results, ...remainingItems],
+      count: firstPage.total
+    };
+  }
+
+  async getRemainingPortalItems(requests) {
+    const pages = await Promise.all(requests);
+    const items = pages.reduce((collection, page) => {
+      return collection.concat(page.results);
+    }, []);
+    return items;
+  }
+
+  getTotalBatch(total, maxPageSize) {
+    return Math.ceil(total / maxPageSize);
+  }
+
+  buildRemainingPageRequests({ portalUrl, query, totalBatch, maxPageSize }) {
+    const requests = [];
+    // Multiple batches each with maxPageSize number of records
+    for (let i = 1; i < totalBatch; i++) {
+      query.start += maxPageSize;
```

Comment:
You may have a problem here.  You are mutating query.start with every iteration of the loop.  I think you will find each request has the same value when the loop is complete. See https://runkit.com/rgwozdz/641dcfbfaed6a40008a63311

Reply:
I have tested it and query.start does increment in each batch accurately. I think this is what might be happening https://runkit.com/sansth1010/641e8d6cf5fdc500084fe2ae . However, I have cloned query object so to not mutate its property.